[1074] ITS deployment

orenyk opened the pull request:
Resolves #1074, finishes setting up Capistrano for deployment.

orenyk commented:
Ok, squashed and ready for review!

mnquintana commented on Gemfile:
```diff
@@ -66,10 +66,11 @@ group :development, :test do
-  gem 'capistrano-bundler', '~> 1.1.4'
-  gem 'capistrano-rails', '~> 1.1.2'
-  gem 'capistrano-rvm', '~> 0.1.2'
+  gem 'capistrano',  '~> 3.3.5', require: false
```
:fire: Extra space here

orenyk replied:
Good catch.

mnquintana commented on config/deploy.rb:
```diff
@@ -9,17 +9,19 @@
-# Set RVM version
-set :rvm_ruby_version, '2.1.2'
+# Set rvm stuff
+set :rvm_ruby_version, File.read('.ruby-version').strip
```
:+1: Awesome

orenyk replied:
Just found out that dotenv-deployment has been deprecated so we should
switch over to dotenv-rails for everything. It shouldn't change anything,
just need to update the Gemfile / `bundle install`.

orenyk commented:
Just updated `dotenv`, moved it to the default Bundler group, and removed `dotenv-deployment` as mentioned in my [last comment](https://github.com/YaleSTC/reservations/pull/1270#discussion_r37668134).

orenyk commented:
We've dealt with a bunch of issues on the ITS side but have finally run into an interesting bug. Since we're updating *very* old databases there are a bunch of migrations to run, including the `EquipmentObject` --> `EquipmentItem` migration followed (later) by the `Reservation.approval_status` removal migration, which does a bunch of processing on existing reservations to convert any information from that column into either `status`, `overdue`, or a tag.

Two issues:
1. Apparently it's really bad practice to use Rails models when touching data in migrations, due to the strong possibility of things getting screwed up later (see [here](http://jacopretorius.net/2014/07/patterns-for-data-migrations-in-rails.html)). Apparently good practice is to write SQL statements directly or create local versions of the model, so we should keep that in mind for the future.
2. More problematically, this migration is failing when it tries to save the reservations since it's running validations, more specifically `matched_item_and_model` which tries to check and make sure that the reservation has been assigned an equipment item, i.e. has been checked out. However, Rails somehow knows that the original association was for an `equipment_object` (even though the model file has changed), so it's trying to read the `equipment_objects` table, which doesn't exist and the migration fails.

Ideally, I wish there were a way to "reload" Rails after running a table rename migration so that all models and associations were up-to-date, but this is also a kludge-y fix for bad code :disappointed:. Take-home number 1 is that we shouldn't rely on Rails models in our migrations anymore. Luckily, this migration calls `save!` at the end so editing to call `save!(validate: false)` shouldn't violate the principle of never editing migrations *too* badly and should resolve the issue in this case. I'll go ahead and do that now.

orenyk commented:
Also, for posterity, we ran into #1282 and I forgot that I had already fixed it so there was temporarily a duplicate fix in this branch. I rebased onto `master` and took the fix from there.

orenyk commented:
Woo hoo, it is fixed! Man, going back to the v3.4.x codebase makes me really appreciate all the work we've done, particularly with `rake db:seed` (thanks @squidgetx!). I've now seeded a database from v3.4.10, had to manually make sure that every reservation with an equipment item had been checked out (since this was not guaranteed and would cause one of our migrations to fail), and then checked out this branch and successfully ran `rake db:migrate`. Hopefully this will be the last bug we have to fix for ITS!

orenyk commented:
We've been asked to add an environment variable to disable *ALL* e-mails for DEV deployments; I'm adding this and using it to optionally set `config.action_mailer.perform_deliveries = false` in `production.rb`.

orenyk commented:
At this point they've got Capistrano deployment working and just have to build the Jenkins wrappers around it, can someone re-review this code so we can merge it into `master`?

orenyk commented:
Ok, @esoterik looked this over and gave it the green light, I'm going to merge it (finally)!
